### PR TITLE
Support a mix of locator and creation configs for SNS

### PR DESCRIPTION
**Pull request by kibertoad**
please review carefully, this was coded in a late night session using the scraps of the mental capacity, likely to be ugly and have gaps.

would be great to clean up the logic and split the code in the follow up prs.

implementation rushed to unblock Billinq service :-/

### CODE REVIEW

File: packages/sns/lib/utils/snsInitter.ts [hunk 1]
```diff
@@ -21,9 +30,9 @@ export async function initSnsSqs(
   extraParams?: ExtraParams,
 ) {
   if (!locatorConfig?.subscriptionArn) {
-    if (!creationConfig?.topic) {
+    if (!creationConfig?.topic && !locatorConfig?.topicArn && !locatorConfig?.topicName) {
       throw new Error(
-        'If locatorConfig.subscriptionArn is not specified, creationConfig.topic parameter is mandatory, as there will be an attempt to create the missing topic',
+        'If locatorConfig.subscriptionArn is not specified, creationConfig.topic, localtorConfig.name or locatorConfig.topicArn parameter is mandatory, as there will be an attempt to create the missing topic',
```
Comment (kamilwylegala) [inline packages/sns/lib/utils/snsInitter.ts:35]: Shouldn't we reword the message that attempt to create topic happens only if `creationConfig` is provided?

File: packages/sns/lib/utils/snsSubscriber.ts [hunk 2]
```diff
@@ -19,14 +24,23 @@ export async function subscribeToTopic(
...
+
+  if (!topicArn) {
+    if (isCreateTopicCommand(topicConfiguration)) {
+      topicArn = await assertTopic(snsClient, topicConfiguration, {
+        queueUrlsWithSubscribePermissionsPrefix:
+          extraParams?.queueUrlsWithSubscribePermissionsPrefix,
+        allowedSourceOwner: extraParams?.allowedSourceOwner,
+      })
+    } else {
+      topicArn = await getTopicArnByName(snsClient, topicConfiguration.topicName)
```
Comment (kamilwylegala) [inline packages/sns/lib/utils/snsSubscriber.ts:41]: If it's not either create command or locate command, shouldn't we throw an error here?

File: packages/sns/lib/sns/AbstractSnsSqsConsumer.ts [hunk 3]
```diff
@@ -85,6 +85,8 @@ export abstract class AbstractSnsSqsConsumer<
         this.creationConfig.queue,
         this.creationConfig.topic,
         this.subscriptionConfig,
+        undefined,
+        this.locatorConfig,
```
Comment (kamilwylegala) [inline packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:89]: TS is complaining now that `initSnsSqs().queueUrl` is string|undefined.

File: packages/sns/lib/sns/AbstractSnsSqsConsumer.ts [hunk 4]
```diff
@@ -12,15 +12,15 @@ import { deleteSnsSqs, initSnsSqs } from '../utils/snsInitter'
...
 
 export type SNSSQSConsumerDependencies = SQSConsumerDependencies & {
   snsClient: SNSClient
 }
 export type SNSSQSCreationConfig = SQSCreationConfig & SNSCreationConfig
 
-export type SNSSQSQueueLocatorType = SQSQueueLocatorType &
-  SNSQueueLocatorType & {
+export type SNSSQSQueueLocatorType = Partial<SQSQueueLocatorType> &
+  SNSTopicLocatorType & {
```
Comment (kamilwylegala) [inline packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:23]: How configuration in billing service could look like?

Arguments passed to constructor should be wrapped with `isTest()` e.g.:
```
{
  "creationConfig": isTest() ? { /* legit creation config */ } : undefined, //In production creation of topics will be disabled.
  "locatorConfig": isTest() ? undefined : { /* legit locate config */ } //In production we will do read-only
}
```

File: packages/sns/lib/utils/snsInitter.ts [hunk 5]
```diff
@@ -68,25 +87,41 @@ export async function initSnsSqs(
...
+    queueName = splitUrl[splitUrl.length - 1]
+  } else {
+    queueName = creationConfig!.queue.QueueName!
+  }
 
   return {
     subscriptionArn: locatorConfig.subscriptionArn,
-    topicArn: locatorConfig.topicArn,
+    topicArn: subscriptionTopicArn,
     queueUrl: locatorConfig.queueUrl,
```
Comment (kamilwylegala) [inline packages/sns/lib/utils/snsInitter.ts:125]: Why does it mix the configuration of both: `locatorConfig` and `creationConfig`? Both can be provided and we return `queueName` from creation config but queue url comes from locate config.

File: packages/sns/lib/utils/snsInitter.ts [hunk 6]
```diff
@@ -97,9 +132,10 @@ export async function deleteSnsSqs(
   snsClient: SNSClient,
   deletionConfig: DeletionConfig,
   queueConfiguration: CreateQueueCommandInput,
-  topicConfiguration: CreateTopicCommandInput,
+  topicConfiguration: CreateTopicCommandInput | undefined,
   subscriptionConfiguration: SNSSubscriptionOptions,
   extraParams?: ExtraParams,
+  topicLocator?: SNSTopicLocatorType,
```
Comment (kamilwylegala) [inline packages/sns/lib/utils/snsInitter.ts:138]: Couldn't we rework this method and introduce a breaking change?

File: packages/sns/lib/utils/snsInitter.ts [hunk 7]
```diff
@@ -150,26 +194,30 @@ export async function deleteSns(
...
-  if (!creationConfig.topic.Name) {
-    throw new Error('topic.Name must be set for automatic deletion')
-  }
+  if (creationConfig.topic) {
+    if (!creationConfig.topic.Name) {
+      throw new Error('topic.Name must be set for automatic deletion')
+    }
 
-  await deleteTopic(snsClient, creationConfig.topic.Name)
+    await deleteTopic(snsClient, creationConfig.topic.Name)
```
Comment (kamilwylegala) [inline packages/sns/lib/utils/snsInitter.ts:202]: Why this change is required?

File: packages/sns/lib/utils/snsInitter.ts [hunk 8]
```diff
@@ -150,26 +194,30 @@ export async function deleteSns(
...
 export async function initSns(
   snsClient: SNSClient,
-  locatorConfig?: SNSQueueLocatorType,
+  locatorConfig?: SNSTopicLocatorType,
   creationConfig?: SNSCreationConfig,
 ) {
   if (locatorConfig) {
-    const checkResult = await getTopicAttributes(snsClient, locatorConfig.topicArn)
+    const topicArn =
+      locatorConfig.topicArn ?? (await getTopicArnByName(snsClient, locatorConfig.topicName))
```
Comment (kamilwylegala) [inline packages/sns/lib/utils/snsInitter.ts:213]: Is it needed? It looks like new API not related to support both creation/locate configs.

**Comment by kamilwylegala** [comment]
EOR1

File: packages/sns/lib/sns/AbstractSnsSqsConsumer.ts (hunk 4 above)
Comment (kibertoad) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:23]: I'd say that creationConfig would always have an entry for the queue, and conditionally will include entry for the topic.
locatorConfig will conditionally either have an entry for the topic, or be empty

File: packages/sns/lib/utils/snsInitter.ts (hunk 1 above)
Comment (kibertoad) [reply packages/sns/lib/utils/snsInitter.ts:35]: yes

File: packages/sns/lib/utils/snsInitter.ts (hunk 5 above)
Comment (kibertoad) [reply packages/sns/lib/utils/snsInitter.ts:125]: I think we only follow this branch in case we ended up using locate config, but I might be wrong, worth checking

File: packages/sns/lib/utils/snsInitter.ts (hunk 6 above)
Comment (kibertoad) [reply packages/sns/lib/utils/snsInitter.ts:138]: potentially. if you see a clear way to do this that would simplify the code, go for it

File: packages/sns/lib/utils/snsInitter.ts (hunk 7 above)
Comment (kibertoad) [reply packages/sns/lib/utils/snsInitter.ts:202]: which part are you referring to?

File: packages/sns/lib/utils/snsInitter.ts (hunk 8 above)
Comment (kibertoad) [reply packages/sns/lib/utils/snsInitter.ts:213]: it is needed to support both having a predefined arn and a lookup by topic name, which new logic relies upon

File: packages/sns/lib/utils/snsSubscriber.ts (hunk 2 above)
Comment (kibertoad) [reply packages/sns/lib/utils/snsSubscriber.ts:41]: yeah, this code is complex and hard to follow. would be nice to break into into more reasonable pieces

File: packages/sns/lib/utils/snsInitter.ts (hunk 1 above)
Comment (kamilwylegala) [reply packages/sns/lib/utils/snsInitter.ts]: Reworded.

File: packages/sns/lib/utils/snsSubscriber.ts (hunk 2 above)
Comment (kamilwylegala) [reply packages/sns/lib/utils/snsSubscriber.ts:41]: I read it again more carefully. Actually it's part of the "locate" logic, but instead of having exact topic arn , we use `topicName` to resolve it into `topicArn`. 

Yes, I believe at least "extract method" could be applied here. I'm on it.

Comment (kamilwylegala) [reply packages/sns/lib/utils/snsSubscriber.ts]: Done, code adjusted.

File: packages/sns/lib/sns/AbstractSnsSqsConsumer.ts (hunk 3 above)
Comment (kamilwylegala) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:89]: @kibertoad I solved this by adding `if` to check if queue url is present. From the code it looks it should be mandatory parameter if `subscriptionArn` is provided, we must provide `queueUrl` too. What do you think?

File: packages/sns/lib/utils/snsInitter.ts (hunk 8 above)
Comment (kamilwylegala) [reply packages/sns/lib/utils/snsInitter.ts:214]: OK, so precedence will always be:
- use ARN if present
- then resolve by name (async lookup)

File: packages/sns/lib/utils/snsInitter.ts (hunk 7 above)
Comment (kamilwylegala) [reply packages/sns/lib/utils/snsInitter.ts:203]: There is extra wrapping condition. I'll simplify it.

File: packages/sns/lib/sns/AbstractSnsSqsConsumer.ts (hunk 4 above)
Comment (kamilwylegala) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:23]: Let me visualize it with the test. I think the rest of the comments will clarify too.

File: packages/sns/lib/sns/AbstractSnsSqsConsumer.ts (hunk 3 above)
Comment (kibertoad) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:89]: isn't there branching after the check you added, where we skip the queueUrl-related logic if it's not present?

we should support passing queueName instead of queueUrl, no? or we always resolve url based on name later?

Comment (kamilwylegala) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:89]: There is branch with check, but it didn't guarantee that `queueUrl` is present after that:
```ts
let queueName: string
  if ((locatorConfig as SQSQueueLocatorType).queueUrl) {
    const splitUrl = (locatorConfig as SQSQueueLocatorType).queueUrl.split('/')
    queueName = splitUrl[splitUrl.length - 1]
  } else {
    queueName = creationConfig!.queue.QueueName!
  }

  return {
    subscriptionArn: locatorConfig.subscriptionArn,
    topicArn: subscriptionTopicArn,
    queueUrl: locatorConfig.queueUrl,
    queueName,
  }
```
From this code it looks that queueUrl is mandatory.

Comment (kibertoad) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:89]: I don't see it. if queueUrl is not present here, we will just return queueUrl: undefined

Comment (kamilwylegala) [reply packages/sns/lib/sns/AbstractSnsSqsConsumer.ts:89]: Yep, but `queueUrl` cannot be null. It's required in `AbstractSnsSqsConsumer`, see line: 104
